### PR TITLE
Distributed datacontract and regex support

### LINKED ISSUE #47: Enhance data contract model to simplify large data contracts
Two additional functionalities that would be good to have
1) Support for external fileDefinitions per file:
currently all fileDefinitions have to be "inline" in the data contract, so large data contracts that support a large number of files get difficult to manage. Support for referencing external files that contain the fileDefinitions model should be added to make management of large data contracts easier
2) Support for regex when comparing input files against fileDefinitions:
Currently only (case-insensitive) string matches are supported. These limit the way file definitions can be managed in a system where input file names might be outside the control of the developer. Adding regex support to the file name matching would be beneficial in this regard.

### PR DESCRIPTION
**Pull request by hammadk373**
Changes:
1) Datacontracts can be reference external json files for fileDefinitions
2) Support regex based matching with input file names. This is controlled by a configuration in the `general` section of the data contract`regexFilenames`. It defaults to false to ensure that legacy behavior is the default.
3) Moved from using python's internal `open(...)` function for reading data contract to using the [smart_open](https://github.com/RaRe-Technologies/smart_open) library. This has the exact same behavior (in fact it internally uses python's `open` function) when dealing with files on the local file system. However, it also add support for cloud storage solutions such as S3, and Azure Object Storage. See doc updates in the PR for more information

Fixes: #47

### CODE REVIEW

File: docs/datacontract.md [hunk 1]
```diff
@@ -41,14 +42,21 @@ conversion process. A single data contract file is used to support multiple CSVT
...
+
 
 
 #### Validations
 - timeZone is a valid value as specified by `pytz.common_timezones`
 - streamType is either historical or live
 
 ### FileDefinition
-The top-level key within a FileDefinition serves as the FileDefinition name. This name is matched against the input CSV file (case-insensitive)
+The top-level key within a FileDefinition serves as the FileDefinition name. This name is matched against the input CSV file using either string match (case-insensitive) or regex (case-sensitive) [see general.regexFilenames setting]
```
Comment (klwhaley) [inline docs/datacontract.md]: nit: can you add periods to complete the sentences in the next two lines

File: docs/datacontract.md [hunk 2]
```diff
@@ -421,3 +439,17 @@ The top-level key within a FileDefinition serves as the FileDefinition name. Thi
 </table>
 
 
+## Alternative Datacontract locations
+CsvToFHIR uses the [smart_open](https://github.com/RaRe-Technologies/smart_open) library to read the Datacontract and any referenced file definitions within it.
```
Comment (klwhaley) [inline docs/datacontract.md:443]: is this now used for all reading of the data contract? even when not in external cloud storage...is that what we want?

Comment (dixonwhitmire) [reply docs/datacontract.md:443]: It looks like the library provides an "open" function which effectively replaces the "open" function from the base library. We can update our setup here to make cloud storage an "opt-in" feature and then check for it's existence at runtime. 

I know that this complicates things a bit, but in the event that we have issues with the library we will still be able to function.

File: tests/model/test_contract.py [hunk 3]
```diff
@@ -122,3 +125,28 @@ def test_file_definition_task_no_required_args(data_contract_data: Dict):
     tasks.append({"name": "set_nan_to_none", "params": None})
     d: DataContract = DataContract(**data_contract_data)
     assert d is not None
+
+
+def test_datacontract_with_invalid_external_config(
```
Comment (klwhaley) [inline tests/model/test_contract.py:130]: are we missing a test for data contract with external config that is successful?

File: docs/datacontract.md (hunk 2 above)
Comment (dixonwhitmire) [reply docs/datacontract.md:443]: Once the other issues, if any, are addressed, I will be happy to submit some changes for this PR to support the "optional" open.

File: tests/model/test_contract.py (hunk 3 above)
Comment (hammadk373) [reply tests/model/test_contract.py:130]: The test for a successful external config is here: https://github.com/LinuxForHealth/CsvToFHIR/blob/distributed_datacontract/tests/test_converter.py#L388. Its a more thorough test that checks both the successful loading and successful conversion of the csv all-in-one. 

Having said that I think a more directed test at the functionality is a good idea. so I added one.

**Comment by hammadk373** [comment]
@klwhaley are there any further changes you want me to look at? 

@dixonwhitmire will you be able to tackle the optional use of the library or do you want me to try and implement that?

**Comment by dixonwhitmire** [comment]
> @klwhaley are there any further changes you want me to look at?
> 
> @dixonwhitmire will you be able to tackle the optional use of the library or do you want me to try and implement that?

Hi @hammadk373 - if @klwhaley feels the issues are addressed I can take a look at the optional use of the library later today. At a high level I was going to update the "setup" to make it an optional component like we've done with jupyter and then use try/except to ensure our "open" call works a-ok with or without the dependency.  If you would prefer to get started on that, that's fine with me. Just let me know.
Thanks!

**Comment by klwhaley** [comment]
I'm going to approve and merge.  And Dixon can do his 'optional' change for the open in another PR.  Then we can cut a release.  Thanks!

**Review by klwhaley — APPROVED** [review]
Approving

**Comment by dixonwhitmire** [comment]
Tracking smart_open as an "optional" component in #50

**Pull request merged**